### PR TITLE
feature(geomean): add geomean function

### LINKED ISSUE #6152: [Feature Request] Add "geometric mean" to categorical ordering
This feature request is to add the geometric mean for ordering/sorting in the categorical axes type.
In particular the addition of `"geomean ascending"` and `"geomean descending"` to `categoryorder` is requested.

The geometric mean is useful since when visualizing/sorting data on a log scale, the arithmetic mean can be very skewed. For larger scales like the log scale having a categorical sort on the geometric mean can provide a better ordering scheme.

Related PRs: https://github.com/plotly/plotly.js/pull/3864

### PR DESCRIPTION
**Pull request by prabhathc**
Add geomean function, closes #6152!

### CODE REVIEW

**Comment by archmoj** [comment]
Thanks very much for the PR.
Would you mind creating a draft log for this PR as described [here](https://github.com/plotly/plotly.js/blob/master/.github/PULL_REQUEST_TEMPLATE.md)?

File: draftlogs/6223_add.md [hunk 1]
```diff
@@ -0,0 +1 @@
+ - Add geomean functionality and 'geomean ascending' + 'geomean descending' to `category_order` on cartesian axes [[#6223](https://github.com/plotly/plotly.js/pull/6223)]
```
[the remark this answers is not in the document]
Comment (acxz) [reply draftlogs/6223_add.md:1]: I am confused as to why `@paulovieira` should be attributed. Did you mean to say `@prabhathc & @acxz`?

[the remark this answers is not in the document]
Comment (archmoj) [reply draftlogs/6223_add.md:1]: Sure. Please update it as required. Thank you too :)

**Comment by nicolaskruchten** [comment]
@alexcjohnson should this be supported in other places like histfunc etc?

**Comment by acxz** [comment]
@nicolaskruchten is that referring to just geomean or the entire categoryorder attribute?

**Comment by nicolaskruchten** [comment]
geomean: all the other aggregation functions you can sort by (sum, mean, median) are also supported in a few other places like `histfunc`, so it would be nice to keep that parallelism going if possible.

**Comment by acxz** [comment]
I'd be down to tackle that in another PR if you can list where else to add the geomean statistic.

**Comment by archmoj** [comment]
Wondering *geomean* may not be the best naming as it could be point to our `geo` subplots.
How do you feel about *geometric mean* or `g-mean` alternatives?
Also there is `harmonic mean` which one may want to add. Right? 
In that case something like `h-mean` works for me; but `harmean` does not :)

**Comment by archmoj** [comment]
On another note, we are wondering why the automatic `CircleCI` tests for this PR are not triggered?!
Do you use any special flags when pushing?

**Comment by acxz** [comment]
yea, `harmean` doesn't sound so nice.
I don't particularly like `g-mean` either since it may not be obvious. I'd actually rather pref `geometric_mean` over `g-mean`.

edit: a quick goog/bing search shows the wikipedia article of Geometric mean with a search query of `g-mean`. I suppose, it is descriptive enough.

**Comment by acxz** [comment]
> On another note, we are wondering why the automatic CircleCI tests for this PR are not triggered?!

I was gonna ask you guys the same thing lol.
There were commits where the CI was triggered, but I believe this stopped when we made this PR a draft and then undid that action.

**Comment by acxz** [comment]
I can try rebasing these commits on master and force pushing to see if that triggers the CI?

**Comment by archmoj** [comment]
> I can try rebasing these commits on master and force pushing to see if that triggers the CI?

Force pushing is discouraged on open pull requests as it could confuse reviewers.
And your branch is only 11 commits behind upstream `master`. So that shouldn't be the reason.
Let's see next time you push a commit into this PR, that may trigger the tests to run on `CircleCI`.

**Comment by archmoj** [comment]
> > On another note, we are wondering why the automatic CircleCI tests for this PR are not triggered?!
> 
> I was gonna ask you guys the same thing lol. There were commits where the CI was triggered, but I believe this stopped when we made this PR a draft and then undid that action.

@antoinerg any idea how we could trigger CI runs on this pull?

**Comment by alexcjohnson** [comment]
> Wondering `geomean` may not be the best naming as it could be point to our geo subplots.

I see your point, though I think in context it's pretty clear. Looking around I see `gmean` used in `scipy.stats` and various npm packages, `geomean` used in MATLAB and Excel, and `geometric_mean` used in the Python stdlib [statistics package](https://docs.python.org/3/library/statistics.html#statistics.geometric_mean). I'm comfortable leaving it as `geomean` but @archmoj if you feel strongly about it I'd be OK with either `gmean` or `geometric mean` (with a space, not a dash or underscore, so the full values would be `geometric mean ascending` or `geometric mean descending`)

Harmonic mean commonly refers to the inverse of the mean of inverses (see eg [npm](https://github.com/stdlib-js/stats-incr-hmean) and [python stdlib](https://docs.python.org/3/library/statistics.html#statistics.harmonic_mean)) so let's stay away from that. But given that we all agree `harmean` is unpleasant, so if we ever did add it we'd need to use `hmean` or `harmonic mean`, perhaps that argues against `geomean` for consistency? In that case I'm leaning toward `geometric mean`, clarity over terseness.

> should this be supported in other places like histfunc etc?

Yes, that would be nice. I only see two more places we could add this, and one of them (the [aggregate transform](https://github.com/plotly/plotly.js/blob/9a2e727f7c30012cd89d63418b880e645eed57ad/src/transforms/aggregate.js#L60)) is deprecated so I'm happy to ignore that. So `histfunc` ([attribute values](https://github.com/plotly/plotly.js/blob/9a2e727f7c30012cd89d63418b880e645eed57ad/src/traces/histogram/attributes.js#L46), [bin functions](https://github.com/plotly/plotly.js/blob/9a2e727f7c30012cd89d63418b880e645eed57ad/src/traces/histogram/bin_functions.js#L22), which get applied [here](https://github.com/plotly/plotly.js/blob/9a2e727f7c30012cd89d63418b880e645eed57ad/src/traces/histogram/calc.js#L118)) is really the only place this needs to be added. @nicolaskruchten am I missing anything?

Two frustrating things though:

(1) we weren't consistent about the naming - for `histfunc` and the aggregate transform we said `avg` rather than `mean`. I would suggest that we add `mean` as a synonym for `avg` and maybe eventually deprecate `avg`, then add `geometric mean` alongside this.

(2) The implementation for `histfunc` is different from that used in category ordering, in that it loops over the array only once, calculating all the quantities it needs for all the bins simultaneously (ie sums and counts), then in a [separate step](https://github.com/plotly/plotly.js/blob/master/src/traces/histogram/average.js) it loops over all the bins to complete the calculation (divide each sum by each count).

File: src/lib/stats.js [hunk 2]
```diff
@@ -47,6 +47,11 @@ exports.mean = function(data, len) {
     return exports.aggNums(function(a, b) { return a + b; }, 0, data) / len;
 };
 
+exports.geoMean = function(data, len) {
+    if(!len) len = exports.len(data);
+    return exports.aggNums(function(a, b) { return a * b; }, 0, data) ** (1 / len);
```
Comment (alexcjohnson) [inline src/lib/stats.js:52]: Does this actually work? I'd think the `0` (the initialization value for the aggregation) needs to be a `1` for geometric mean. Also you probably need to include whatever we name this in the [regex that separates the aggregator from the order](https://github.com/plotly/plotly.js/blob/a10b927bd64259d029578e3f5948ded2275c2ac5/src/plots/plots.js#L3095)

**Comment by archmoj** [comment]
I'd vote for `geometric mean` too, when there is an election here :)

File: src/lib/stats.js (hunk 2 above)
Comment (archmoj) [reply src/lib/stats.js:52]: On another note `**` operator is not working on IE according to the following:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Exponentiation

Perhaps better to replace it `Math.pow`?

**Comment by nicolaskruchten** [comment]
> Two frustrating things though

Ah yeah I remember these. Well, probably reason enough not to do extra work here then :)

File: draftlogs/6223_add.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (acxz) [reply draftlogs/6223_add.md]: Resolved

File: src/lib/stats.js (hunk 2 above)
Comment (acxz) [reply src/lib/stats.js]: > I'd think the 0 (the initialization value for the aggregation) needs to be a 1 for geometric mean.
include whatever we name this in the [regex that separates the aggregator from the order](https://github.com/plotly/plotly.js/blob/a10b927bd64259d029578e3f5948ded2275c2ac5/src/plots/plots.js#L3095)

> ** operator is not working on IE

Great catches
resolved

**Comment by acxz** [comment]
>  Let's see next time you push a commit into this PR, that may trigger the tests to run on CircleCI.

I have already tried that.

> `geometric mean`

done.

**Comment by acxz** [comment]
> how we could trigger CI runs on this pull?

For right now, I'll just make a duplicate pull to see the tests and be able to iterate on them if they are incorrect.

**Comment by archmoj** [comment]
I fetched your branch and pushed it to plotly.js repository at the `geomean`branch. 
See https://github.com/plotly/plotly.js/compare/geomean
The tests are now running at https://app.circleci.com/pipelines/github/plotly/plotly.js/6863/workflows/f4c7b53b-1569-4ba3-b492-e46418c1b39d.

**Comment by archmoj** [comment]
With the changes `npm run test-jasmine calcdata` should pass.
Would you please have a look at the failing tests here?
https://app.circleci.com/pipelines/github/plotly/plotly.js/6866/workflows/7690f800-a50d-4321-97bb-1e19d7814ec8/jobs/156536/parallel-runs/6?filterBy=FAILED

Also if you fetch `upstream/master` and merge it into this branch the "export test" would be successful.

**Comment by acxz** [comment]
Ah okay sweet.

**Comment by acxz** [comment]
@archmoj I'm not sure how to fix the following error: https://app.circleci.com/pipelines/github/plotly/plotly.js/6866/workflows/7690f800-a50d-4321-97bb-1e19d7814ec8/jobs/156536/parallel-runs/6?filterBy=FAILED&invite=true#step-104-1026

**Comment by archmoj** [comment]
@prabhathc Could you please fetch `upstream/master` and merge `master` into this branch to resolve the conflicts?

**Comment by acxz** [comment]
@archmoj done, merge caused a name change from `geometric mean` to `geometric_mean`. Please confirm if that is the api we want to go with.

File: test/jasmine/tests/calcdata_test.js [hunk 3]
```diff
@@ -1160,6 +1160,23 @@ describe('calculated data and points', function() {
                         checkAggregatedValue(baseMock, expectedAgg, false, done);
                     });
 
+                    it('takes the geometric mean of all values per category across traces of type ' + trace.type, function(done) {
```
Comment (archmoj) [inline test/jasmine/tests/calcdata_test.js:1163]: Let's skip `ohlc` and `candlestick` which fail with tiny tolerance.
```suggestion
                    it('takes the geometric mean of all values per category across traces of type ' + trace.type, function(done) {
                        if(trace.type === 'ohlc' || trace.type === 'candlestick') return done();

```

**Comment by archmoj** [comment]
To reflect the changes to the API, please run 
```sh
npm run schema && git add test/plot-schema.json && git commit -m "update plot-schema diff"
```

File: test/jasmine/tests/calcdata_test.js (hunk 3 above)
Comment (acxz) [reply test/jasmine/tests/calcdata_test.js:1163]: Sure thing.

**Comment by acxz** [comment]
@archmoj done, all the tests finally pass!

**Review by archmoj — APPROVED** [review]
Nicely done.
:dancer:
Thank you!

**Pull request merged**